src/transforms/exclusive_route/config.rs

```diff
@@ -0,0 +1,179 @@
+use vector_lib::config::clone_input_definitions;
+
+pub(super) const UNMATCHED_ROUTE: &str = "_unmatched";
+
+/// Individual route configuration.
+#[configurable_component]
+#[derive(Clone, Debug)]
+pub struct Route {
+    /// The name of the route is also the name of the transform port.
+    ///
+    ///  The `_unmatched` name is reserved and thus cannot be used as route ID.
+    ///
+    /// Each route can then be referenced as an input by other components with the name
+    ///  `<transform_name>.<name>`. If an event doesn’t match any route,
+    /// it is sent to the `<transform_name>._unmatched` output.
+    pub name: String,
+
+    /// Each condition represents a filter which is applied to each event.
+    pub condition: AnyCondition,
+}
+
+impl Hash for Route {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.name.hash(state);
+    }
+}
+
+impl PartialEq for Route {
+    fn eq(&self, other: &Self) -> bool {
+        self.name == other.name
+    }
+}
+
+impl Eq for Route {}
+
+/// Configuration for the `route` transform.
+#[configurable_component(transform(
+    "exclusive_route",
+    "Split a stream of events into multiple sub-streams based on user-supplied conditions."
+))]
+#[derive(Clone, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct ExclusiveRouteConfig {
+    /// An array of named routes. The route names are expected to be unique.
+    #[configurable(metadata(docs::examples = "routes_example"))]
+    pub routes: Vec<Route>,
+}
+
+fn routes_example() -> Vec<Route> {
+    vec![
+        Route {
+            name: "foo-and-bar-exist".to_owned(),
+            condition: AnyCondition::Map(ConditionConfig::Vrl(VrlConfig {
+                source: "exists(.foo) && exists(.bar)".to_owned(),
+                ..Default::default()
+            })),
+        },
+        Route {
+            name: "only-foo-exists".to_owned(),
+            condition: AnyCondition::Map(ConditionConfig::Vrl(VrlConfig {
+                source: "exists(.foo)".to_owned(),
+                ..Default::default()
+            })),
+        },
+    ]
+}
+
+impl GenerateConfig for ExclusiveRouteConfig {
+    fn generate_config() -> toml::Value {
+        toml::Value::try_from(Self {
+            routes: routes_example(),
+        })
+        .unwrap()
+    }
+}
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "exclusive_route")]
+impl TransformConfig for ExclusiveRouteConfig {
+    async fn build(&self, context: &TransformContext) -> crate::Result<Transform> {
+        let route = ExclusiveRoute::new(self, context)?;
+        Ok(Transform::synchronous(route))
+    }
+
+    fn input(&self) -> Input {
+        Input::all()
+    }
+
+    fn validate(&self, _: &schema::Definition) -> Result<(), Vec<String>> {
+        let set: HashSet<Route> = self.routes.iter().cloned().collect();
+        let mut errors = Vec::new();
+
+        if self.routes.len() != set.len() {
+            errors.push("Found routes with identical names.".into());
+        }
```

Comment:
It'd be useful to output the duplicate names. I suppose you could iterate over `set` and check how many occurrences are in `self.routes` with `iter().filter(...).count() > 1`. Slow (`n^2`), but this check is at start up and so should be fine.